### PR TITLE
Refactoring of multi-topic kafka ingestion docs

### PR DESCRIPTION
**Pull request by abhishekagarwal87**
In this PR, I have gotten rid of `multiTopic` parameter and instead added a `topicPattern` parameter. Kafka supervisor will pass `topicPattern` or `topic` as the stream name to the core ingestion engine. There is validation to ensure that only one of topic or topicPattern will be set. This new setting is easier to understand than overloading the `topic` field that earlier could be interpreted differently depending on the value of some other field.

### CODE REVIEW

File: docs/development/extensions-core/kafka-supervisor-reference.md [hunk 1]
```diff
@@ -144,6 +144,12 @@ impressions, you will set `topic` to `clicks|impressions` in the IO config. If n
 match the regex, druid will automatically start ingesting from those new topics. If you enable multi-topic
```
Comment (vogievetsky) [inline docs/development/extensions-core/kafka-supervisor-reference.md:144]: Capitalize `Druid` as a proper noun

File: docs/development/extensions-core/kafka-supervisor-reference.md [hunk 2]
```diff
@@ -37,7 +37,7 @@ This topic contains configuration reference information for the Apache Kafka sup
 
 |Field|Type|Description|Required|
 |-----|----|-----------|--------|
-|`topic`|String|The Kafka topic to read from. Must be a specific topic. Topic patterns are not supported.|yes|
+|`topic`|String|The Kafka topic to read from. Must be a specific topic. Topic patterns are supported if `multiTopic` is enabled.|yes|
```
Comment (vogievetsky) [inline docs/development/extensions-core/kafka-supervisor-reference.md:40]: I think this should be restructured slightly to a form like this:

```
If `multiTopic` is not set (or false) then `topic` is treated as a string that represents the Kafka topic from which the data will be read.
If `multiTopic` is true then `topic` represents a regular expression that can match (exactly) several topics.
```

One of the things I am not clear on (and want to see documentation for) is the `exactly` part.
Is the topic regex anchored in the start and end?
If I have topics like `metrics-stage`, `metrics-dev`, `metrics-prod` if multiTopic is true should topic be set to `metrics-` or `metrics-.*`  ?

**Comment by vogievetsky** [comment]
Please also check out https://github.com/apache/druid/pull/14833

File: docs/development/extensions-core/kafka-supervisor-reference.md (hunk 2 above)
Comment (abhishekagarwal87) [reply docs/development/extensions-core/kafka-supervisor-reference.md:40]: rephrasing it a bit more 
The `topic` is interpreted differently depending on whether the multi-topic ingestion (link-to-anchor for multi-topic) is enabled or not. If multi-topic ingestion is enabled, `topic` represents a regular expression that can match several topics. Check (link-to-anchor) for more details. If multi-topic ingestion is not enabled, topic is considered a single kafka topic to read from.

Comment (abhishekagarwal87) [reply docs/development/extensions-core/kafka-supervisor-reference.md]: I have made the above changes.

File: docs/development/extensions-core/kafka-supervisor-reference.md [hunk 3]
```diff
@@ -37,7 +37,8 @@ This topic contains configuration reference information for the Apache Kafka sup
 
 |Field|Type|Description|Required|
 |-----|----|-----------|--------|
-|`topic`|String|The Kafka topic to read from. Must be a specific topic. Topic patterns are not supported.|yes|
+|`topic`|String|The Kafka topic to read from. Must be a specific topic. Use this setting when you want to ingest from a single kafka topic.|yes|
```
Comment (kfaraz) [inline docs/development/extensions-core/kafka-supervisor-reference.md:40]: Nit: caps
```suggestion
|`topic`|String|The Kafka topic to read from. Must be a specific topic. Use this setting when you want to ingest from a single Kafka topic.|yes|
```

File: docs/development/extensions-core/kafka-supervisor-reference.md [hunk 4]
```diff
@@ -138,11 +138,20 @@ The following example demonstrates supervisor spec with `lagBased` autoScaler an
...
+
+To ingest data from multiple topics, you have to set `topicPattern` in the supervisor IO config and not set `topic`.
+Multiple topics can be passed as a regex pattern as the value for `topicPattern` in the IO config. For example, to
+ingest data from clicks and impressions, you will set `topicPattern` to `clicks|impressions` in the IO config.
+Similarly, you can use `metrics-.*` as the value for `topicPattern` if you want to ingest from all the topics that
+start with `metrics-`. If new topics are added to the cluster that match the regex, Druid will automatically start
+ingesting from those new topics. If you enable multi-topic ingestion for a datasource, downgrading to a version
+lesser than 28.0.0 will cause the ingestion for that datasource to fail.
+
+When ingesting data from multiple topics, the partitions are assigned based on the hashcode of topic and the id of the
```
Comment (kfaraz) [inline docs/development/extensions-core/kafka-supervisor-reference.md:150]:
```suggestion
When ingesting data from multiple topics, the partitions are assigned based on the hashcode of the topic name and the id of the 
```

**Review by kfaraz — APPROVED** [review]
Seems better than having a separate boolean and easier to understand. 🚀

**Pull request merged**